### DISCUSSION TITLE
track? preserve? for variables dervied from other parameters

### DISCUSSION THREAD

**Discussion by oralb** [cropbox/Cropbox.jl#14] [Q&A] [answered]
I am constructing a system. Below is the brief code;

@system test(Controller) begin
---omitted upper parts---
      
    a: a ~ preserve(parameter)
    b: b ~ preserve(parameter)
  
    alpha(a, b, c): factor => begin
        (a + b) 
    end ~ preserve? track?

--- omitted lower parts-----
 end

My question is about the "alpha" in the code. The alpha is derived from a and b which are "preserve(parameter)". 
I believe that alpha should also be "preserve(parameter)". But "track" is OK when the system was constructed and then performed.

I wonder which one is close to the developer's intention and Cropbox's performance.

Thanks in advance!

**Comment by tomyun** [comment, marked as answer]
The choice between `preserve` and `track` boils down to intended behaviors of the variable that is declared. Generally, the most of `preserve` variables are used as parameters, but then there are still some less common cases where we want non-parameter `preserve` variables, e.g., automatically derived parameter or static assignment of external value. You can then create a `preserve` variable with computational logic, but `track` also should be able to do the same work as you mentioned.

In such case, I'd recommend using `track`. There is a chance that its depending parameters later replaced by some dynamic variables after systems composition, leading to an unexpected logical error hard to be caught up.

Technically, `preserve` has a performance advantage when compared to `track` since the former only generates code that runs once with no further update. Practically, the actual difference should be slim as we currently have major bottlenecks in other parts of the framework.
